truffle/src/main/ruby/core/string.rb

```diff
@@ -1208,16 +1208,26 @@ def scrub(replace = nil, &block)
     end
 
     taint = tainted?
+
+    validate = -> str {
+      str = StringValue(str)
+      unless str.valid_encoding?
+        raise ArgumentError, "replacement must be valid byte sequence"
+      end
+      # Special encoding check for #scrub
+      if str.ascii_only? ? !encoding.ascii_compatible? : encoding != str.encoding
+        raise Encoding::CompatibilityError, "incompatible character encodings"
+      end
+      taint = true if str.tainted?
```

Comment:
I don't really care for modifying the captured value like this. It's too easy to read as setting a local. Maybe add a comment?